### DISCUSSION TITLE
Rule ordering is nondeterministic in development server with `important: true`

### DISCUSSION THREAD

**Discussion by phoenixeliot** [tailwindlabs/tailwindcss#7625] [Help]
**What version of Tailwind CSS are you using?**
v3.0.23

**What build tool (or framework if it abstracts the build tool) are you using?**
Next.js 12.1.0

**What version of Node.js are you using?**

v16.13.2

**What browser are you using?**

Brave

**What operating system are you using?**

macOS

**Reproduction URL**

https://github.com/phoenixeliot/bug-tailwindcss-nondeterministic

**Describe your issue**

When using the development server and `important: true`, the applied CSS order is nondeterministic (changes based on order of edits made to the code).

Reproduction steps:
- Clone repo, `yarn`, `yarn run dev`
- Open index.js, inspect the element
- Add a `!` before `rounded-full`, observe that CSS changes in the inspector but display doesn't change
- Remove the `!`, observe how CSS changes again (with duplicates now) and the page display changes
- Restart the dev server and reload the page, see the display has gone back to the original

<img width="924" alt="Screen Shot 2022-02-22 at 14 12 35" src="https://user-images.githubusercontent.com/491376/155202334-f117a1db-208a-4b5b-a2e4-d6c693d1232e.png">
<img width="874" alt="Screen Shot 2022-02-22 at 14 12 45" src="https://user-images.githubusercontent.com/491376/155202340-aee5bfd5-bc70-49a1-a3f1-e78ff2a14281.png">
<img width="871" alt="Screen Shot 2022-02-22 at 14 12 53" src="https://user-images.githubusercontent.com/491376/155202350-d2097e66-7b76-403a-a07b-76f6dc29d00e.png">

I understand that one shouldn't apply conflicting rules, but this is a problem we're running into when trying to create a component library, when we want to override any of the component's rules.

**Comment by adamwathan** [comment]
Hey! Yeah ordering is not deterministic for colliding rules, even without the `!important` stuff. For example `p-4` and `p-6` could be in your stylesheet in any order, but `px-4` will always appear after `p-4`.

Can you explain in more detail what makes this an actual problem in your situation? We show a lint warning with our IntelliSense extension when you add conflicting classes to an element because we really do just consider it a user error.

**Comment by phoenixeliot** [comment]
The problem in our codebase was trying to have a base React component, `<Button>` which had (among others) `rounded py-3 px-10 text-base`.
Then we wanted to occasionally be able to customize it to be a different size in another context, so we pass in a `className` and append that to the className in the JSX. If we pass in rules that collide with the ones already in Button (since it's hard to replace, much easier to add competing class names), I was finding that some worked and some didn't, and if I used for example `!rounded-lg` in an attempt to make it take precedence, then it would appear to have worked... but then if I restart the dev server, or changed the code in another order, the CSS order would change and my code would render differently than it had before restarting. This means there's an ever-present hazard that we might develop something that looks fine when we're writing it, but then when deployed via production build, looks significantly different. This means I have to be paranoid about whether what I wrote really will produce what I'm looking at, when customizing components.

This makes customizing components much harder, and basically forces adding a lot of prop options to internally manage which sets of classnames get applied. Which can work, but it's a bit of a heavy solution.

**Comment by phoenixeliot** [comment]
The problem of rules getting duplicated in the stylesheet seems to only occur with `important: true`; if I don't have that on, rules don't get duplicated (though ! rules which get added afterward do get added in the order you type them in your code)

I'm not sure if this is deterministic or just based on the order they're seen when parsing the files, or something.

**Comment by adamwathan** [comment]
Yeah so the ordering is always non-deterministic and based on the order of your edits/the order that your content files are parsed, behaves the same way regardless of whether `important: true` is set, or at least that's the intention.

The thing is that for your case, even if it was deterministic it wouldn't matter right? If you had `rounded` and you wanted to override it with `rounded-full` maybe that would work, but now if another component has `rounded-full` and you wanted to override that with `rounded` _that_ wouldn't work — only one of those can work.

I'm going to convert this to a discussion since this isn't a bug from our perspective and` is by design, and the real issue you're trying to solve here is "how can I add two classes to an element that are in conflict and have control over which class wins?", which is totally a valid question and one that comes up a lot.

What do you think of a solution like this instead?

```html
<div class="initial:bg-blue-500 bg-green-500"></div>
```

The `initial` modifier reduces the specificity of the class selector to `001` so that the `010` class selector will override it. Here's a working demo with a simple `initial` variant created in user-land:

https://play.tailwindcss.com/ohmUlyBeOT

**Reply by bill-kerr** [reply]
I wrote a plugin along the lines of what Adam is proposing above. Maybe it'll help you out in this case.

https://github.com/bill-kerr/tailwindcss-plugin-defaults

**Reply by adamwathan** [reply]
@bill-kerr Nice! Your plugin could be a lot simpler if you change the variant to this:

`addVariant('initial', 'html :where(&)')`

Then you won't need to replace all of Preflight, because the `initial:bg-black` utilities will have a high enough specificity to override those rules thanks to the `html` at the beginning.

**Reply by bill-kerr** [reply]
Well I guess that makes a lot of sense. Thanks for pointing it out 😃

Looking forward to something like this being included in the base package!

**Reply by kungpaogao** [reply]
love this idea - what's the current "official" recommended solution for dealing with ordering/composing classes for components?

edit: related thread here https://github.com/tailwindlabs/tailwindcss/discussions/1446

**Comment by simonc** [comment]
Hi there.

One thing that's somewhat bothering with the classes order being non-deterministic is that is also makes for bigger diff than necessary in PRs. It makes it a bit harder to see what was really added vs. what has simply changed position in the file.

Also I can confirmed the issue with `important: true` and classes being duplicated. For some reason some of them seem to not be cleared when no more needed and for instance a case that was troubling was:

* Adding a custom maxWidth in my Tailwind config
* Applying it to an element
* Keep working for a moment
* Later on, temporarily removing the custom maxWidth while keeping the class on the element
* Being puzzled by the style still being applied

Cutting the `watch` and running a ` build` cleared things up. I don't remember the real customized property but it was less obvious than a maxWidth 😅

I wouldn't say those are actual bugs but IMHO they do create unexpected/confusing behaviors.

My 2 cents ❤️
